WIP: draft of CIP for "CC Fee Removal" (#2)
meiersi-da/cips: cip-CC-FEE-REMOVAL/cip-CC-FEE-REMOVAL.md

<pre>
  CIP:  CIP XXXX
  Title: Canton Coin Fee Removal
  Author:
    Simon Meier
  License: CC0-1.0
  Status: Working Draft
  Type: Tokenomics
  Created: 2025-08-14
  Approved:
</pre>

## Abstract

TODO: write

> ## Motivation

Reviewer: iirc putting motivation first is inconsistent with what cip 0 specifies. I personally do prefer it but I think Wayne wanted to stick to cip 0.

Author: done. I also like the motivation first, but not worth fighting over it.

The motivation is as follows:

* Removing CC fees for `TransferPreapproval` contracts simplifies user setup.
  In particular, it allows users to create a preapproval for receiving CC before they own CC.
  The traffic cost of setting up a `TransferPreapproval` is significant enough to prevent abuse,
  so there is no need for an additional CC fee.

* Switching to direct collection of validator rewards simplifies user setup, ensures traffic purchasers
  get rewarded for their activity, and solves the one outstanding security issue (CC-3)
  identified in the [Quantstamp security audit of Canton Coin](https://lists.sync.global/g/tokenomics/message/575).

  User setup is simplified because no `ValidatorRight` contracts need to be set up
  to designate the validator operator party for a user.
  Not having to manage `ValidatorRight` contracts is especially useful when migrating between validator operators,
  or when [recovering CC balance from keys only](https://docs.dev.sync.global/validator_operator/validator_disaster_recovery.html#re-onboard-a-validator-and-recover-balances-of-all-users-it-hosts).

  Switching from indirect to direct collection of validator rewards is

>   desired because setting the CC fees to zero means that all validator activity

Reviewer: I think also cns burns? or do we not issue a validator reward for that?

Author: we use an `AmuletRules_Transfer` as part of the payment collection, which means that validator reward will be set to zero. 

this cip does not change how CNS works. Not clear yet whether we'll issue any rewards for using CNS once we adjust it. I suspect not.

Reviewer: makes sense, thanks!

## Specification

The changes in this CIP are initially implemented as changes to the Amulet Rules configuration where possible,
and changes to the Splice code where necessary.
In the future, the Splice code might be further simplified by completely removing the logic for handling CC fees.

### Remove Canton Coin Fees

Change the Amulet Rules configuration parameters as follows:

| Configuration                               | Value     |
|---------------------------------------------|-----------|
| `transferConfig.createFee.fee`              | `0.0`     |
| `transferConfig.transferFee.initialRate`    | `0.0`     |
| `transferConfig.transferFee.steps.0._2`     | `0.0`     |
| `transferConfig.transferFee.steps.1._2`     | `0.0`     |
| `transferConfig.transferFee.steps.2._2`     | `0.0`     |
| `transferConfig.lockHolderFee.fee`          | `0.0`     |

### Adjust Holding Fees

Change the Daml code for Amulet such that no holding fees are charged when using a coin
as an input to a transfer. Combined with the config changes to remove CC fees this
guarantees that the sum of coin inputs is always equal to the sum of coin outputs in a transfer.

> Holding fees are still charged when a coin contract is expired by the DSO

Reviewer: This sounds a bit like holding fees are charged in addition to the coin getting expired. Maybe rephrase it as something like "SVs retain the option to expire coins after a given amount of time dependent on the value but no holding fees are charged when a coin is used in a transfer.

because the holding fees surpassed the coin's initial amount, which ensures
that "dust coins" (see [Rationale](#rationale)) do not accumulate and slow down the system.

### Switch to Direct Validator Reward Collection

Change the Daml code for Amulet such that validator activity records can
only be used to mint rewards by the user that generated them.
Deprecate `ValidatorRight` contracts and no longer require them to collect validator rewards.

> Change the automatic reward collection mechanism in the Splice Wallet to select

Reviewer: This sounds slightly confusing given that the logic right now queries activity records for all users for which a validator right exists. You could argue those are "user's validator activity records". Maybe something like "to select only the user's own validator activity records instead of selecting the validator activity records for all users on the node when minting for the operator party"

Author: good point. fixed.

<pre>
  CIP:  CIP XXXX
  Title: Canton Coin Fee Removal
  Author:
    Simon Meier
  License: CC0-1.0
  Status: Draft
  Type: Tokenomics
  Created: 2025-09-01
  Approved:
</pre>

## Abstract

This CIP proposes to remove Canton Coin fees and adjust holding fees so that
users do not have to pay fees when transferring Canton Coin and application
developers can build applications without special code to deal with these fees.
Thus making Canton Coin more attractive for both users and application

> developers.

Reviewer 2: ```suggestion
This will make Canton Coin, and the Global Synchronizer, more attractive for both users and application
developers.
```

### Adjust holding Fees

Change the Daml code for Amulet such that no holding fees are charged when using a coin
as an input to a transfer. Combined with the config changes to remove CC fees this
guarantees that the sum of coin inputs is always equal to the sum of coin outputs in a transfer.

SVs retain the ability to expire coins after a time period proportional to the value.
Concretely, they can continue to exercise the `Amulet_Expire` choice
when the holding fees for a coin surpass the initial amount of the coin. These
holding fees continue to be determined as before this CIP using the formula:
```
holdingFees currentRound (ExpiringAmount initialAmount createdAtRound ratePerRound) =
  ratePerRound * (currentRound - createdAt)
```
where the `ExpiringAmount` parameters are stored on the `Amulet` contract and are set
as part of the transfer that created the `Amulet` contract.

### Adjust Splice App UIs

Change UIs to reflect the removal of Canton Coin fees and the adjustment of holding fees.
Concretely, this means:

- Remove references to Canton Coin fee parameters from Splice App UIs,
  except for UIs that show the internal configuration of the Amulet Rules.

> - Ensure the UI elements only show non-zero fee values.

Reviewer 2: what does this mean?

Reviewer: In the scan and SV UI, where you we currently display the fee values, hide them when they are zero to not confuse users once we set it to zero.

Reviewer 2: Still not exactly clear what this means; maybe a screenshot illustrating your intent? 

I would have expected something like:
* Remove all Canton Coin fees, and explanatory text, from the "Fees" section of the Canton Coin Scan UI:
<img width="622" height="787" alt="image" src="https://github.com/user-attachments/assets/ba9b3af0-6df5-46e9-a6fb-77e60cc21251" />

* Remove the "Current Canton Coin Configuration" header and explanatory text from the same UI
* Reformat and highlight the "Synchronizer Fee" line, since it is the only aspect of the Coin configuration that will remain in this portion of the screen. 

I'm not sure where else fees appear in the UI, other than in the DSO config pages, which should stay the same as today.

- Do not deduct holding fees from totals or coin balances in Splice App UIs.
  The only place where holding fees should be shown is in the transaction details
  for a transaction that expires a "dust coin".
- Test that there are no division-by-zero problems due to the fee parameters
  being set to zero.

## Motivation

This CIP proposes to remove Canton Coin fees and adjust holding fees so
that users do not have to pay fees when transferring Canton Coin and
application developers can build applications without special code to

> deal with these fees. Thus making Canton Coin more attractive for both users and

Reviewer 2: ```suggestion
deal with these fees. This will make Canton Coin more attractive for both users and
```

  User setup is simplified because no `ValidatorRight` contracts need to be set up
  to designate the validator operator party for a user.
  Not having to manage `ValidatorRight` contracts is especially useful when migrating between validator operators,
  or when [recovering CC balance from keys only](https://docs.dev.sync.global/validator_operator/validator_disaster_recovery.html#re-onboard-a-validator-and-recover-balances-of-all-users-it-hosts).

  Switching from indirect to direct collection of validator rewards is
  desired because setting the CC fees to zero means that all validator activity
  records are due to traffic purchases. Direct collection thus results in the right tokenomics:
  the purchaser gets rewarded instead of the validator operator party of the
  validator node hosting the purchaser.

## Rationale

### Burn on MainNet

To substantiate the claim that the majority of burn on MainNet is due to traffic purchases,
we can look at the burn statistics for MainNet.
In the 3 month time period from 2025-04-01 to 2025-07-31,
the total burn on MainNet was 446'680'130 CC out of which
423'475'365 CC were due to traffic purchases. Thus about

> 94.8% of the burn was due to traffic purchases.

Author: @stephencompall-DA: can you do me a favor and check that the dates and numbers line up with what you measured when running your script? Thanks a lot.

Reviewer 3: They don't; it was run with [these parameters](https://github.com/hyperledger-labs/splice/blob/85ce9e1fd5c9ecf107a132a95e83a15cd0e57947/apps/app/src/test/resources/traffic-cost-burn-time-period.sql#L115-L118). With 4-1 to 7-1 you get

```json
[{
  "purchase_burns": "220544190.6744",
  "non_purchase_burns": "11532873.1989568527"
}]
```

which is 95%.

Reviewer 3: 4-1 to 7-31 would be a 4-month time period.

> In the 3 month time period from 2025-04-01 to 2025-07-31,
> the total burn on MainNet was 446'680'130 CC out of which
> 423'475'365 CC were due to traffic purchases. Thus about
> 94.8% of the burn was due to traffic purchases.

Author: ```suggestion
In the 3 month time period from 2025-05-14 to 2025-08-14,
the total burn on MainNet was 446'680'130 CC out of which
423'475'365 CC were due to traffic purchases. Thus about
94.8% of the burn was due to traffic purchases.
```
@stephencompall-DA : ah I've overlooked the timestamps in your script. I've added them now to the suggestion. Could you double-check that one again? thx :blush:

Reviewer 3: 👍

<pre>
  CIP:  CIP XXXX

>   Title: Canton Coin Fee Removal

Author: Should probably be "Canton Coin Usage Fees Removal"

instead of indirectly via the validator operator party
and to adjust the CC fees for creating or renewing `TransferPreapproval` contracts so
that short-lived preapprovals (expiry < 90 days) can be created without paying CC fees.
Thus making Canton Coin more attractive for both users and application
developers.

## Specification

The changes in this CIP are initially implemented as changes to the Amulet Rules configuration where possible,
and changes to the Splice code where necessary.
In the future, the Splice code might be further simplified by completely removing the logic for handling CC fees.

### Remove Canton Coin Fees

Change the Amulet Rules configuration parameters as follows:

| Configuration                               | Value     |
|---------------------------------------------|-----------|
| `transferConfig.createFee.fee`              | `0.0`     |
| `transferConfig.transferFee.initialRate`    | `0.0`     |
| `transferConfig.transferFee.steps.0._2`     | `0.0`     |
| `transferConfig.transferFee.steps.1._2`     | `0.0`     |
| `transferConfig.transferFee.steps.2._2`     | `0.0`     |
| `transferConfig.lockHolderFee.fee`          | `0.0`     |

> 

Author: Missing: note that Daml must be changed such that reward issuance on a transfer grants the `extraFeaturedAppRewardAmount` even when no usage fees have been paid.

<pre>
  CIP:  CIP XXXX
  Title: Canton Coin Fee Removal
  Author:
    Simon Meier
  License: CC0-1.0
  Status: Draft
  Type: Tokenomics
  Created: 2025-08-19
  Approved:
</pre>

## Abstract

This CIP proposes to remove Canton Coin fees and adjust holding fees so that
users do not have to pay fees when transferring Canton Coin and application
developers can build applications without special code to deal with these fees.

> This CIP further proposes to have users collect their validator rewards directly

Author: collect -> mint

> This CIP further proposes to have users collect their validator rewards directly

Author: remove from abstract

Author: formulate it as technical cleanup

Author: state this as a positive assertion that users that purchase traffic will get credit for the activity. This is a related change that's now possible because we are removing usage fees.

instead of indirectly via the validator operator party
and to adjust the CC fees for creating or renewing `TransferPreapproval` contracts so
that short-lived preapprovals (expiry < 90 days) can be created without paying CC fees.
Thus making Canton Coin more attractive for both users and application

> developers.

Author: Note to reviewers: given recent discussions, it might be better to remove these two changes and land them as part of a later CIP.

Happy to do so.

## Specification

The changes in this CIP are initially implemented as changes to the Amulet Rules configuration where possible,
and changes to the Splice code where necessary.
In the future, the Splice code might be further simplified by completely removing the logic for handling CC fees.

### Remove Canton Coin Fees

Change the Amulet Rules configuration parameters as follows:

| Configuration                               | Value     |
|---------------------------------------------|-----------|
| `transferConfig.createFee.fee`              | `0.0`     |
| `transferConfig.transferFee.initialRate`    | `0.0`     |
| `transferConfig.transferFee.steps.0._2`     | `0.0`     |
| `transferConfig.transferFee.steps.1._2`     | `0.0`     |
| `transferConfig.transferFee.steps.2._2`     | `0.0`     |
| `transferConfig.lockHolderFee.fee`          | `0.0`     |

Change the Daml code for rewards issuance in `AmuletRules_Transfer` such that featured app rewards

> are issued independently of whether CC usage fees were charged.

Reviewer 3: Does this need to, or accidentally, affect the other choices with an embedded transfer? (i.e. these perform a transfer but do not have a descendant `AmuletRules_Transfer` exercise)

* `AmuletRules_CreateExternalPartySetupProposal`
* `AmuletRules_CreateTransferPreapproval`
* `AmuletRules_BuyMemberTraffic`
* `TransferPreapproval_Renew`

Author: The Daml PR for this changes the one place where rewards are issued, and also fixes these calls you list here. 

I'll fix the explanation here to also cover them.

Author: @moritzkiefer-da : I've rewritten the text to explain the `ValidatorRight` change in terms of solving a design mistake wrt traffic purchases.

However, when trying to resolve this comment I noticed that the choices @stephencompall-DA mentions will actually still produce `ValidatorRewardCoupon`s and they are thus not only about traffic purchases.

Do you have a good idea for the more general name / functionality that we issue validator reward activity records for?

i.e. what's the commonality between all the choices that use `splitAndBurn`?

Side note for @stephencompall-DA : none of these choices issue app rewards. So they are not affected by the change wrt featured app rewards that is mentioned in the text here.

Author: change to validator activity record minting removed from text
